File: python/combinations.py
#
# combinations.py - generate lexicographic combinations of symbols
#
import time

#
# Algorithm Algorithm 7.2.1.3T TAOCP by Donald Knuth: Lexicographic Combinations
# This algorithm is like Algorithm L, but faster. It assumes, for convenience, t < n
#
# Return a list of digits from 1 to n taken t at a time, in lexicographic order
#
def combine(n, t):
  # T1. Initialize
  c = [i for i in range(t)]
  c.append(n)   # c[t] = n
  c.append(0)   # c[t+1] = 0
  j = t
  x = 0

  def findj():
    nonlocal x, j, c
    while (True):
      c[j-2] = j - 2
      x = c[j-1] + 1
      if x != c[j]:
        return j
      j = j + 1

  while j <= t:
    # T2. Visit. At this point j is the smallest index such that c[j+1] > j
    # Visit the combination ct ... c2 c1
    result = [c[i] for i in range(t)]

    if j > 0:
      x = j
      # T4. Increase c[j]
      c[j-1] = x
      j = j - 1
    # T3. Easy case?
    elif (c[0] + 1) < c[1]:
      c[0] = c[0] + 1
    else:
      j = 2
      j = findj()
      # T5. Terminate the algorithm if j > t
      if j <= t:
        c[j-1] = x
        j = j - 1
    yield result

for combination in combine(6, 3):
  print(combination)

millis = int(round(time.time() * 1000))
count = 0
for comb in combine(100, 5):
  count = count +1
now = int(round(time.time() * 1000))
print("Took %3.3f seconds to generate %d results" % ((now - millis) / 1000.0, count))

File: C/combinations.c
#include "combinations.h"
#include <stdlib.h>

#ifndef COMBINATIONS_H
#define COMBINATIONS_H

int has_next(combination_t *self) {
	return self->j <= self->t;
}

void visit(combination_t *self, int *buffer) {
	for (int i=0; i < self->t; i++) {
		buffer[i] = self->c[i];
	}
}

void findj(combination_t *self) {
	int *c = self->c;
	while (1) {
		int j = self->j;
		c[j-2] = j - 2;
		self->x = c[j-1] + 1;
		if (self->x != c[j]) {
			return;
		}
		self->j++;
	}
}

void next(combination_t *self, int *buffer) {
	visit(self, buffer);

	int j = self->j;
	int *c = self->c;

	if (j > 0) {
		self->x = j;
		// T4. Increase c[j]
		c[j-1] = self->x;
		self->j--;
	}
	// T3. Easy case?
	else if (c[0] + 1 < c[1]) {
		c[0]++;
	}
	else {
		self->j = 2;
		findj(self);
		// T5. Terminate the algorithm if j > t
		if (self->j <= self->t) {
			// T6. Increase c[j]
			c[self->j-1] = self->x;
			self->j--;
		}
	}
}

combination_t *new_combination(int n, int k) {
	combination_t *self = malloc(sizeof(combination_t));
	self->n = n;
	self->t = k;
	int *c = malloc(sizeof(int) * (n + 2));
	for (int i=0; i < k; i++) {
		c[i] = i;
	}
	c[k] = n;
	c[k+1] = 0;
	self->c = c;
	self->j = k;
	return self;
}

#endif